src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs

```diff
@@ -5,313 +5,117 @@ namespace Microsoft.DocAsCode.Build.Engine
-                        };
-
-                        RemoveDuplicateOutputFiles(generatedManifest.Files);
-
-                        // post process
-                        foreach (var postProcessor in postProcessors)
-                        {
-                            using (new LoggerPhaseScope(postProcessor.Item1))
-                            {
-                                generatedManifest = postProcessor.Item2.Process(generatedManifest, parameters.OutputBaseDir);
-                                if (generatedManifest == null)
-                                {
-                                    throw new DocfxException($"Plugin {postProcessor.Item1} should not return null manifest");
-                                }
-
-                                // To make sure post processor won't generate duplicate output files
-                                RemoveDuplicateOutputFiles(generatedManifest.Files);
-                            }
-                        }
-
-                        _manifest.AddRange(generatedManifest.Files);
-                        _homepages.AddRange(generatedManifest.Homepages);
-                        _xrefMaps.Add((string)generatedManifest.XRefMap);
-
-                        // Last step: save manifest file
-                        var manifestJsonPath = Path.Combine(parameters.OutputBaseDir, Constants.ManifestFileName);
-                        JsonUtility.Serialize(manifestJsonPath, generatedManifest);
-                        Logger.LogInfo($"Manifest file saved to {manifestJsonPath}.");
-
-                        Logger.LogInfo($"Completed building {generatedManifest.Files?.Count} file(s).");
-                    }
-                }
-                finally
-                {
-                    if (hostServices != null)
-                    {
-                        foreach (var item in hostServices)
-                        {
-                            Cleanup(item);
-                            item.Dispose();
-                        }
+                        throw new DocfxException($"Plugin {postProcessor.ContractName} should not return null metadata");
                     }
                 }
             }
         }
 
-        private void UpdateUidDependency(DocumentBuildContext context, List<HostService> hostServices)
+        public void PostProcess(Manifest manifest, string outputDir)
         {
-            foreach (var hostService in hostServices)
+            // post process
+            foreach (var postProcessor in _postProcessors)
             {
-                foreach (var m in hostService.Models)
+                using (new LoggerPhaseScope(postProcessor.ContractName))
                 {
-                    if (m.Type == DocumentType.Overwrite)
-                    {
-                        continue;
-                    }
-                    if (m.LinkToUids.Count == 0)
+                    manifest = postProcessor.Processor.Process(manifest, outputDir);
```

Comment:
Line 157 seems have done that.

---
In reply to: [71098498](https://github.com/dotnet/docfx/pull/501#discussion_r71098498) [](ancestors = 71098498)